tests/frontend/test_program.py

```diff
@@ -672,6 +672,81 @@ class DummyCircuit(Compiler):
                 compiler=DummyCircuit(),
             )
 
+    def test_missing_layout(self):
+        """Test that an error is raised if the device spec is missing a layout when compiled"""
+        device_dict = {
+            "target": None,
+            "layout": None,
+            "modes": 2,
+            "compiler": ["DummyCompiler"],
+            "gate_parameters": {
+                "squeezing_amplitude_0": [0, 1],
+            },
+        }
+
+        class DummyCircuit(Compiler):
+            """A circuit with 2 modes"""
+
+            interactive = True
+            primitives = {"S2gate"}
+            decompositions = set()
+
+        spec = sf.DeviceSpec(spec=device_dict)
+
+        prog = sf.Program(2)
+        with prog.context as q:
+            ops.S2gate(0.6) | [q[0], q[1]]
+
+        with pytest.raises(ValueError, match="missing a circuit layout"):
+            new_prog = prog.compile(
+                device=spec,
+                compiler=DummyCircuit(),
+            )
+
+    def test_missing_gate_parameters(self):
+        """Test that an error is raised if the device spec is missing gate parameters when compiled"""
+        mock_layout = textwrap.dedent(
+            """\
+            name mock
+            version 1.0
+
+            S2gate({squeezing_amplitude_0}, 0.0) | [0, 1]
+            """
+        )
+
+        device_dict = {
+            "target": None,
+            "layout": mock_layout,
+            "modes": 2,
+            "compiler": ["DummyCompiler"],
+            "gate_parameters": None,  # no gate_parameters, so any value is valid
```

Comment:
Should we add a test for the case where `gate_parameters` is an empty dictionary? This could happen if a new hardware device is registered which is not configurable.